semrel.go

```diff
@@ -211,84 +214,117 @@ func (listRefsQuery) vars(repo *Repository, perPage int, cursor string) map[stri
-				break REFS
-			} else if verRangeConstraint != nil && verRangeConstraint.Check(r.Version) {
-				// if verRange is set and matches this release, we are done
-				return r, nil
-			} else if rPreRel, _, _ := strings.Cut(r.Version.Prerelease(), "."); rPreRel == prerelease {
-				// If the last production release was newer, we already would have
-				// stopped, so we know this is newer.
-				if prerelease != "" {
-					if verRangeConstraint == nil {
-						return r, nil
-					}
-					lastRelease = r
-					break REFS
+			n++
+		}
+		log.Println("Skipped", n, "non-release versions")
+		return nil
+	})
+	var lastPreRelease *Release
+	if prerelease != "" {
+		// locate pre-release
+		eg.Go(func() error {
+			log.Println("Searching for pre-release version:", prerelease)
+			n := 0
+			for r, err := range repo.tags(prerelease) {
+				// this will often find some false positives that we need to skip over
+				if err != nil {
+					return err
+				} else if rPreRel, _, _ := strings.Cut(r.Version.Prerelease(), "."); rPreRel == prerelease {
+					log.Println("Found latest matching pre-release version:", r.Version.String())
+					lastPreRelease = r
+					break
 				}
+				n++
 			}
-		}
+			log.Println("Skipped", n, "non-matching pre-release versions")
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
 
-		if !q.Repository.Refs.PageInfo.HasNextPage {
-			break
+	if prerelease == "" {
+		return lastMainRelease, nil
+	}
+
+	if lastPreRelease != nil {
+		// if we found both a pre-release and default branch release tag, we need to
+		// see which is newer
+		if lastMainRelease == nil {
+			return lastPreRelease, nil
+		} else if lastPreRelease.Version.GreaterThan(lastMainRelease.Version) {
+			log.Println("Last pre-release version is higher than last main release")
+			return lastPreRelease, nil
 		}
-		cursor = string(q.Repository.Refs.PageInfo.EndCursor)
 	}
+	return lastMainRelease, nil
+}
 
-	// pretend the latest release is from verRange, applying any pre-release tag it has
-	if verRangeVer, err := semver.NewVersion(verRange); err != nil {
-		return nil, fmt.Errorf("unable to parse maintained version %q as semver: %w", verRange, err)
-	} else if _, verRangePre, vrHasPre := strings.Cut(verRange, "-"); !vrHasPre {
-		// verRange is just a version, no pre-release segment. Attach its version to
-		// the latest release's commit
-		return &Release{lastRelease.SHA, verRangeVer}, nil
-	} else if verWithPreRel, err := verRangeVer.SetPrerelease(verRangePre); err != nil {
-		return nil, fmt.Errorf("failed to initialize pre-release for mainted version %q: %w", verRange, err)
-	} else {
-		// verRange is a version and a pre-release tag. Attach this combo to the
-		// latest release's commit
-		return &Release{lastRelease.SHA, &verWithPreRel}, nil
+// tags generates a sequence of Release objects built from tags that match the
+// given query string (which may be empty) in reverse chronological order by
```

Comment:
What ensures the ordering here.. is that just how the GQL interface returns things?

Reply:
The GraphQL query includes specific ordering parameters, see line 200

Reply:
Ah righto